### PR TITLE
feat: Add span_preprocessor hook

### PR DESCRIPTION
**Pull request by a1flecke**
- allow net::http spans to be preprocessed before sending

### CODE REVIEW

File: instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb
Comment (ericmustin) [inline instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb:26]: This change is in contradiction with the opentelemetry specification for HTTP Span Names. https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/http/#name

> HTTP client spans have no http.route attribute since client-side instrumentation is not generally aware of the “route”, and therefore HTTP client spans SHOULD use {http.method}. Instrumentation MUST NOT default to using URI path as span name, but MAY provide hooks to allow custom logic to override the default span name.

So we can't accept this change upstream.

**Comment by a1flecke** [comment]
> In this case, we don't have a hook on this instrumentation at the moment, so I think the recommended approach would be to use an opentelemetry-collector processor. Would that be a viable path forward for you?

@ericmustin Thank you for this information.  With our current setup a processor is not viable.  Is there a way that I can help to add a hook to the HTTP client instrumentation?

**Comment by ericmustin** [comment]
@a1flecke We'd need to implement a hook, I don't believe we have an example of this for http client spans, but this example(from rack) would probably look a bit similar to whatever would need to be added to http client instrumentation https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/55f52c461d552bc5e97f9832751d3fa6853e9b03/instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb#L162-L172

I can bring this up at the SIG mtg today and see if there's other ideas or a possible workaround.

**Comment by a1flecke** [comment]
@ericmustin I took a stab at adding a hook to the http client.  Please let me know what you think

**Review by williamsbdev — APPROVED** [review]

**Comment by ericmustin** [comment]
@williamsbdev thanks for the quick turnaround here. Awesome stuff, the updates here you've included as definitely along the right path.  However, rather than an arbitrary span hook, ideally the callable would take specific fields/strings relevant to the request, and returns a string that is assigned explicitly as the span name. So for example it could take the `method` and `path` as arguments, or ideally any/all of the below

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/55f52c461d552bc5e97f9832751d3fa6853e9b03/instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb#L18-L23

Basically the idea is we intentionally don't want to expose a hook that would allow arbitrary modification to the span, and we also don't want to expose a hook that would allow arbitrary modifications to the request object. However, we do want to permit a hook that accepts specific known arguments and returns a string that is explicitly set as the span name.

There's a bit of backstory here but tl;dr exposing arbitrary hooks pre/post processing hooks becomes a nightmare to maintain and debug and we don't want to support that at the moment.

I would use the rack example i've linked to above as a guide. if any @open-telemetry/ruby-contrib-approvers want to weigh in here on which request fields would be appropriate to pass in as arguments to this callable, please do so. i think for your use case _just_ method + path is enough?

**Comment by a1flecke** [comment]
@ericmustin took another crack at it.  Current approach only allows the span name to be updated and restricts which attributes are sent to the callable

**Review by arielvalentin — CHANGES_REQUESTED** [review]
@a1flecke please sign the Easy CLA at your convenience.

**Comment by ericmustin** [comment]
@a1flecke hey, appreciate the second attempt here. still a bit off from what i was hoping for, thats my fault i did a bad job of communicating. After thinking this through, what i think is acceptable according to the specification, and what has the least performance tradeoffs, is that the lambda shouldn't modify span name and shouldn't access the span attributes (it means acquiring a lock on the mutex every time we touch the span name or it's attributes, iirc, so we want to limit that especially on a hot path like http client spans). The least contentious lambda config option here is one that only accepts method and path as arguments, that seems to be what the specification allows and is also what the rack example does, and explicitly sets that as the span name at span creation time (rather than updating the span name).

i've pushed up a few changes to this branch that reflect this (_my local env is a mess so bear with me as i work thru linting/test fails on CI_), wdyt? does this help unblock y'all or is it too restrictive?

Also @a1flecke please sign the CLA when you have a moment.

Also @williamsbdev i hear you on the yolo approvals, your enthusiasm for this config option is duly noted.

**Comment by a1flecke** [comment]
@ericmustin  looks good.  I signed the CLA.  Thanks for the assist

**Comment by a1flecke** [comment]
@ericmustin @arielvalentin is there a way to retrigger the GH action status checks?  I believe they failed due to the GH incident that happened today: https://www.githubstatus.com/incidents/nf7s6933tnn8

**Review by ericmustin — APPROVED** [review]
i'm going to add some safety to these callables in a followup pr, footgun waiting to happen

**Review by ericmustin — APPROVED** [review]

File: instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb
Comment (a1flecke) [reply instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb:50]: As a user of this library, the current name does not make sense to me. I am also fairly new to OTLP. I am deferring to the maintainers that the name chosen makes sense in the OTLP domain. 

I will also admit that I am biased towards getting this change in and released. I am about to add OTLP to many rails applications and really want more control over the generated span names

Comment (a1flecke) [reply instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb:50]: @arielvalentin The name you proposed or the name from my earlier commit was clearer to me. Again, I trust those who are in the domain more than me.

Comment (ericmustin) [reply instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb:50]: I have no strong opinions here I just was following the style of rack's url_quantization. Happy to update to @arielvalentin suggestion

Comment (ericmustin) [reply instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb]: yea that makes sense, updated

**Comment by a1flecke** [comment]
@ericmustin @arielvalentin Thanks for all of your help on this.  This looks good to me.  Anything else to do before it is merged?

File: instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb
Comment (arielvalentin) [inline instrumentation/http/lib/opentelemetry/instrumentation/http/patches/client.rb:49]: This should probably have some error handling and fall back to `HTTP #{request_method}` in the case where an error occurs.

**Comment by a1flecke** [comment]
@arielvalentin is anything else needed?

**Review by arielvalentin — APPROVED** [review]

**Comment by a1flecke** [comment]
@arielvalentin @ericmustin I just want to say thank you again for your help and quick feedback on this PR.  It was a pleasure to contribute to this project.

**Comment by a1flecke** [comment]
@arielvalentin @ericmustin can you provide guidance on how often this project releases?  I did not a clear pattern in the Github releases page.

**Comment by a1flecke** [comment]
> @a1flecke I usually look to release after the weekly SIG on Tuesdays or on-demand depending on the severity. E.g. for bug fixes I open release requests as soon as I merge.

@arielvalentin Can you please create a release with this change?

**Comment by a1flecke** [comment]
nvm, I found it: https://github.com/open-telemetry/opentelemetry-ruby-contrib/releases/tag/opentelemetry-instrumentation-http%2Fv0.23.0

Thanks!

**Pull request merged**
